File: app/src/main/java/uz/kidzone/app/BackendClient.java
package uz.kidzone.app;

import android.util.Log;
import com.google.firebase.auth.FirebaseAuth;
import java.io.IOException;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public final class BackendClient {

    private static final String TAG = "BackendClient";
    private static final String BASE_URL = "https://kidzone-backend-s7to.onrender.com";

    private BackendClient() {}

    public static void sendTopicPush(String title, String body, String url,
                                     Runnable onDone, Runnable onError) {
        if (FirebaseAuth.getInstance().getCurrentUser() == null) {
            Log.w(TAG, "sendTopicPush: no Firebase user");
            if (onError != null) onError.run();
            return;
        }
        FirebaseAuth.getInstance().getCurrentUser()
            .getIdToken(false)
            .addOnSuccessListener(result -> doPost(result.getToken(), title, body, url, onDone, onError))
            .addOnFailureListener(e -> {
                Log.w(TAG, "getIdToken failed: " + e);
                if (onError != null) onError.run();
            });
    }

    private static void doPost(String idToken, String title, String body, String url,
                                Runnable onDone, Runnable onError) {
        try {
            org.json.JSONObject data = new org.json.JSONObject();
            data.put("url", url != null ? url : "");
            org.json.JSONObject payload = new org.json.JSONObject();
            payload.put("title", title != null ? title : "");
            payload.put("body", body != null ? body : "");
            payload.put("data", data);

            RequestBody reqBody = RequestBody.create(
                payload.toString(), MediaType.parse("application/json"));
            Request request = new Request.Builder()
                .url(BASE_URL + "/push/send-all")
                .addHeader("Authorization", "Bearer " + idToken)
                .post(reqBody)
                .build();

            KidZoneApplication.getHttpClient().newCall(request).enqueue(new Callback() {
                @Override public void onFailure(Call call, IOException e) {
                    Log.w(TAG, "sendTopicPush failed: " + e);
                    if (onError != null) onError.run();
                }
                @Override public void onResponse(Call call, Response response) {
                    Log.d(TAG, "sendTopicPush HTTP " + response.code());
                    response.close();
                    if (response.isSuccessful()) { if (onDone != null) onDone.run(); }
                    else { if (onError != null) onError.run(); }
                }
            });
        } catch (org.json.JSONException e) {
            Log.w(TAG, "JSON error: " + e);
            if (onError != null) onError.run();
        }
    }
}

File: app/src/main/java/uz/kidzone/app/BackendClient.kt
package uz.kidzone.app

import android.util.Log
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

object BackendClient {

    private const val TAG = "BackendClient"
    private const val BASE_URL = "https://kidzone-backend-s7to.onrender.com"

    /** Registers the FCM token with the backend so UID-targeted pushes work. Failure is logged only. */
    @JvmStatic
    fun registerToken(uid: String, fcmToken: String) {
        val currentUser = com.google.firebase.auth.FirebaseAuth.getInstance().currentUser
        if (currentUser == null) {
            Log.w(TAG, "registerToken: no Firebase user")
            return
        }
        currentUser.getIdToken(false)
            .addOnSuccessListener { result ->
                doRegisterToken(result.token, fcmToken)
            }
            .addOnFailureListener { e ->
                Log.w(TAG, "registerToken: getIdToken failed: $e")
            }
    }

    private fun doRegisterToken(idToken: String?, fcmToken: String) {
        try {
            val payload = JSONObject().apply {
                put("fcmToken", fcmToken)
            }
            val mediaType = "application/json".toMediaType()
            val reqBody = payload.toString().toRequestBody(mediaType)
            val request = Request.Builder()
                .url("$BASE_URL/push/register-token")
                .addHeader("Authorization", "Bearer $idToken")
                .post(reqBody)
                .build()

            KidZoneApplication.httpClient.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.w(TAG, "registerToken: HTTP failed: $e")
                }

                override fun onResponse(call: Call, response: Response) {
                    Log.d(TAG, "registerToken HTTP ${response.code}")
                    response.close()
                }
            })
        } catch (e: JSONException) {
            Log.w(TAG, "registerToken: JSON error: $e")
        }
    }

    /** Fetches the first active banner from the backend. Calls callback(null) on any error. */
    @JvmStatic
    fun fetchFirstActiveBanner(callback: (BannerChecker.BannerData?) -> Unit) {
        val request = Request.Builder()
            .url("$BASE_URL/banners")
            .get()
            .build()

        KidZoneApplication.httpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.w(TAG, "fetchFirstActiveBanner: HTTP failed: $e")
                callback(null)
            }

            override fun onResponse(call: Call, response: Response) {
                try {
                    if (!response.isSuccessful) {
                        Log.w(TAG, "fetchFirstActiveBanner: HTTP ${response.code}")
                        response.close()
                        callback(null)
                        return
                    }
                    val bodyStr = response.body?.string() ?: ""
                    response.close()
                    val arr = org.json.JSONArray(bodyStr)
                    if (arr.length() == 0) { callback(null); return }
                    val first = arr.getJSONObject(0)
                    val title = first.optString("title", "")
                    val desc  = first.optString("description", "")
                    val link  = first.optString("link", "")
                    if (link.isEmpty()) { callback(null); return }
                    callback(BannerChecker.BannerData(title = title, body = desc, url = link))
                } catch (e: Exception) {
                    Log.w(TAG, "fetchFirstActiveBanner: parse error: $e")
                    callback(null)
                }
            }
        })
    }
}
